chunky/src/java/se/llbit/chunky/renderer/scene/SkyCache.java

```diff
@@ -106,44 +119,42 @@ public Vector3 calcIncidentLight(Ray ray) {
 
   // Linear interpolation between 2 points in 1 dimension
   private double interp1D(double x, double x0, double x1, double y0, double y1) {
-    return y0 + (x - x0)*(y1-y0)/(x1-x0);
+    return y0 + (x - x0) * (y1 - y0) / (x1 - x0);
   }
 
-  // Calculate the bilinearly interpolated value from the cache.
+  /**
+   * Calculate the bilinearly interpolated value from the cache.
+   */
   private Vector3 getColorInterpolated(double normX, double normY) {
     double x = normX * skyResolution;
     double y = normY * skyResolution;
-    int floorX = (int) QuickMath.clamp(x, 0, skyResolution-1);
-    int floorY = (int) QuickMath.clamp(y, 0, skyResolution-1);
-
-    if (skyTexture[floorX][floorY][0] < 0) bake(floorX, floorY);
-    if (skyTexture[floorX][floorY+1][0] < 0) bake(floorX, floorY+1);
-    if (skyTexture[floorX+1][floorY][0] < 0) bake(floorX+1, floorY);
-    if (skyTexture[floorX+1][floorY+1][0] < 0) bake(floorX+1, floorY+1);
+    int floorX = (int) QuickMath.clamp(x, 0, skyResolution - 1);
+    int floorY = (int) QuickMath.clamp(y, 0, skyResolution - 1);
 
     double[] color = new double[3];
     for (int i = 0; i < 3; i++) {
-      double y0 = interp1D(x, floorX, floorX + 1, skyTexture[floorX][floorY][i], skyTexture[floorX+1][floorY][i]);
-      double y1 = interp1D(x, floorX, floorX + 1, skyTexture[floorX][floorY+1][i], skyTexture[floorX+1][floorY+1][i]);
+      double y0 = interp1D(x, floorX, floorX + 1, skyTexture[floorX][floorY][i],
+          skyTexture[floorX + 1][floorY][i]);
+      double y1 = interp1D(x, floorX, floorX + 1, skyTexture[floorX][floorY + 1][i],
+          skyTexture[floorX + 1][floorY + 1][i]);
       color[i] = interp1D(y, floorY, floorY + 1, y0, y1);
     }
-
     return new Vector3(color[0], color[1], color[2]);
   }
 
-  // Calculate the sky color for a pixel on the cache.
-  private void bake(int x, int y) {
+  /**
+   * Calculate the sky color for a pixel on the cache.
+   */
+  private Vector3 getSkyColorAt(int x, int y) {
     Ray ray = new Ray();
 
     double theta = ((double) x / skyResolution) * 2 * PI;
-    double phi = ((double) y / skyResolution) * PI - PI/2;
+    double phi = ((double) y / skyResolution) * PI - PI / 2;
     double r = FastMath.cos(phi);
     ray.d.set(FastMath.cos(theta) * r, FastMath.sin(phi), FastMath.sin(theta) * r);
 
     Vector3 color = simSky.calcIncidentLight(ray);
     ColorUtil.RGBtoHSL(color, color.x, color.y, color.z);
-    skyTexture[x][y][0] = color.x;
-    skyTexture[x][y][1] = color.y;
-    skyTexture[x][y][2] = color.z;
+    return color;
```

Comment:
By not writing to `skyTexture` here, this method doesn't need to be synchronized anymore and can be used concurrently.